pkg/services/searchV2/manager.go

```diff
@@ -0,0 +1,299 @@
+				// Index for org already exists, do nothing.
+				m.mu.RUnlock()
+				close(signal.done)
+				continue
+			}
+			m.mu.RUnlock()
+			lastIndexedEventID := lastEventID
+			// Prevent full re-indexing while we are building index for new org.
+			// Full re-indexing will be later re-started in `case lastIndexedEventID := <-reIndexDoneCh`
+			// branch.
+			fullReIndexTimer.Stop()
+			go func() {
+				// We need semaphore here since asynchronous re-indexing may be in progress already.
+				asyncReIndexSemaphore <- struct{}{}
+				defer func() { <-asyncReIndexSemaphore }()
+				err = m.buildInitialIndex(ctx, signal.orgID)
+				signal.done <- err
+				reIndexDoneCh <- lastIndexedEventID
+			}()
+		case <-fullReIndexTimer.C:
+			// Periodically rebuild indexes since we could miss updates. At this moment we are issuing
+			// entity events non-atomically (outside of transaction) and do not cover all possible dashboard
+			// change places, so periodic re-indexing fixes possibly broken state. But ideally we should
+			// come to an approach which does not require periodic re-indexing at all. One possible way
+			// is to use DB triggers, see https://github.com/grafana/grafana/pull/47712.
+			lastIndexedEventID := lastEventID
+			go func() {
+				// Do full re-index asynchronously to avoid blocking index synchronization
+				// on read for a long time.
+
+				// We need semaphore here since re-indexing due to build signal may be in progress already.
+				asyncReIndexSemaphore <- struct{}{}
+				defer func() { <-asyncReIndexSemaphore }()
+
+				started := time.Now()
+				m.logger.Info("Start re-indexing")
+				err := m.reIndexExisting(ctx)
+				if err != nil {
+					m.logger.Error("Full re-indexing finished with error", "fullReIndexElapsed", time.Since(started), "error", err)
+				} else {
+					m.logger.Info("Full re-indexing finished", "fullReIndexElapsed", time.Since(started))
+				}
+				reIndexDoneCh <- lastIndexedEventID
+			}()
+		case lastIndexedEventID := <-reIndexDoneCh:
+			// Asynchronous re-indexing is finished. Set lastEventID to the value which
+			// was actual at the re-indexing start – so that we could re-apply all the
+			// events happened during async index build process and make sure it's consistent.
+			if lastEventID != lastIndexedEventID {
+				m.logger.Info("Re-apply event ID to last indexed", "currentEventID", lastEventID, "lastIndexedEventID", lastIndexedEventID)
+				lastEventID = lastIndexedEventID
+				// Apply events immediately.
+				partialUpdateTimer.Reset(0)
+			}
+			fullReIndexTimer.Reset(reIndexInterval)
+		case <-ctx.Done():
+			return ctx.Err()
+		}
+	}
+}
+
+func (m *orgIndexManager) buildInitialIndexes(ctx context.Context, orgIDs []int64) error {
+	started := time.Now()
+	m.logger.Info("Start building in-memory indexes")
+	for _, orgID := range orgIDs {
+		err := m.buildInitialIndex(ctx, orgID)
+		if err != nil {
+			return fmt.Errorf("can't build initial dashboard search index for org %d: %w", orgID, err)
```

Comment:
```suggestion
			return fmt.Errorf("can't build initial %s search index for org %d: %w", m.name, orgID, err)
```

also nit, but it would be nice if we had `w.name` (+`orgId` when possible) in all logs/errors

Reply:
Done, changed in all logs in file